design/rest-api-guideline.md

```diff
@@ -0,0 +1,389 @@
+
+#### Group
+
+The `API_GROUP` refers to a set of one or more functionality that can be grouped together. The endpoints associated with a group can manage different resources as far as they are strictly related with the functionality represented by the group.
+
+By including the group in the resource path, we aim to provide a clear and consistency view of the functionalities exposed by the different APIs, with more control over the groups of functionality (in terms of functional and non-functional aspects) as well as make more clear how we communicate the functional structure of these APIs through the documentation.
+
+Here is an example of a resource path related with the group of functionalities provided by the Account Management REST API:
+
+```
+/apis/accounts/v1/{user_id}
+```
+
+The path refers to the `accounts` group and any resource/operation related with user account management (in this case for user with id `user_id`) can be accessed from that path.
+Note that by grouping functionalities together and in conjunction with the versioning schema, we can keep the root path and also the semantics of the API without impact the adaptability for new changes or improvements to the API.
+
+Group names should be in lowercase, a single word name and plural.
+
+#### Versioning
+
+Versioning helps APIs to adapt to changes to their fields or restructure resource representations. In addition to that, versioning can help developers to offer non-stable, technology preview features as well as control the end-of-life of APIs.
+
+The `version` refers to the stability and the state of an API. The version does not necessarily refers to a Keycloak release version number so that it can be mapped to different Keycloak release versions.
+
+Here is an example of a resource path related with the group of functionalities provided by the Account Management REST API:
+
+```
+/apis/accounts/v1/{user_id}
+```
+
+From the example above we can infer that the path refers to a specific `v1` version with the possibility to change the version and: 
+
+* completely change the API
+* do breaking changes to parts of the API
+* remove functionalities or specific parts of the API
+* provide non-stable or technology preview features
+* the different APIs can evolve separately
+
+#### Extensions
+
+To better acommodate and separate the core API from extensions built on top of the API, developers should consider the following path format for extensions:
+
+```
+/apis/{API_GROUP}/extensions/{version}
+```
+
+This separation aims to avoid confusion when consuming stable and official Keycloak APIs so that clients can differentiate if they are using some API provided by a third-party with no support from Keycloak team.
+
+#### Technology Preview
+
+To better acommodate and separate the core API from technology preview features, developers should use a beta version as the API version:
```

Comment:
The proposed format also makes clear that the API is not supported. The work to change a root path in an application should quite trivial. Note that this is all about the root path. The sub-resources would still be the same when we move a beta API into our list of supported APIs.

Reply:
Does it mean that once we change our support level for given version (i.e. we decide that some preview versions is now good enough to be supported), the old API just stop working?

Reply:
Yeah, but just like any deprecation policy, we will give some time to clients to migrate.

Reply:
The name should be v1preview1 or something, not beta1. There is a fundamental difference between the two when it comes to support. As long as the API is in preview we should not provide any deprecation policies, but simply remove and replace. This should be made clear to people though.

Reply:
True. So I guess we can avoid versioning in this case and just use `/preview/{api}` to communicate a tech preview API.

Reply:
@pedroigor That means the "preview" part of the path would be in place of a version? So instead of `/v1/users/{id}` would be something like `/preview/users/{id}`?
If so, +1 :)

Reply:
+1

Reply:
Yes. Like this `/realms/{realm}/apis/{API_GROUP}/{version}/preview/{preview_id}` or `/realms/{realm}/apis/preview/{preview_id}`.

Reply:
@pedroigor Actually, what I meant was that `preview` is the version identification. So no `{preview_id}` should be required. Because, if I understand it correctly, there will be at most only one preview version available at the same time. And there will be no deprecation policy - when a feature is not a preview anymore, it's just gone from the `preview` version.

In other words, instead of `v1`, `v2` and similar, you would use `preview` (instead of the actual version).

Reply:
The `preview_id` is a name of the preview future. In your example, it would be `users`. This is similar to the extension format.

Reply:
@pedroigor I'm still not sure I fully understand what `preview_id` should be. It seems to me a bit confusing, especially when we use `version` at the same time. Could you give me some specific real-life example how the full API path would look like, what `preview_id` could actually be?

Aren't previews often used in place of a version, something like `/{realm}/apis/accounts/v1preview1/{user_id}`?

Reply:
There are two scenarios here.

Consider you are adding a new feature set that you want to bake in the community. If the feature set is a new API Group you will have a path like:

```
/{realm}/apis/preview/newfeatureset
```

Now, if the feature set is added to an existing API Group such as `accounts`, you will have something like:

```
/{realm}/apis/accounts/v1/preview/newfeatureset
```

Is it more clear now? See that we decided to only version API Groups (e.g.: accounts).

Reply:
@pedroigor Thanks for clarification! So, if I understand it correctly, we won't use the preview "labeling" when modifying existing feature sets? It'll be applicable only for new feature sets that didn't exist before?

Say, we want to remove some field from user accounts resource. Would we somehow mark such change as a preview?

Reply:
In this case, if the change can break backward compatibility, I think best is have a new version of `accounts` API Group. It does not fit into preview as we are actually modifying the API and not including a tech preview feature.

Reply:
@pedroigor Ah, I think I understand now! Thank you for your patience. :)
Technical Preview is actually not related to versioning. It's a tech preview in the same sense as we have in tech previews in Keycloak - it's just a separated feature set. Am I correct this time? If so, consider this resolved.